appcues/src/main/java/com/appcues/action/ActionProcessor.kt

```diff
@@ -26,9 +26,21 @@ internal class ActionProcessor(override val scope: Scope) : KoinScopeComponent {
         }
     }
 
-    fun process(action: ExperienceAction) {
+    fun process(actions: List<ExperienceAction>) {
         appcuesCoroutineScope.launch {
-            actionQueue.send(action)
+            transformQueue(actions).forEach {
+                actionQueue.send(it)
+            }
+        }
+    }
+
+    private fun transformQueue(actions: List<ExperienceAction>) = actions.fold(actions) { currentQueue, action ->
+        val indexInCurrent = currentQueue.indexOfFirst { it === action }
+        val transformingAction = action as? ExperienceActionQueueTransforming
+        if (indexInCurrent != -1 && transformingAction != null) {
+            transformingAction.transformQueue(currentQueue, indexInCurrent, appcues)
+        } else {
+            currentQueue
```

Comment:
learned about `fold` - its much like the `reduce` used in Swift - it walks over each item in the list and allows an accumulator function to run.  In our case, this accumulator is either just returning the queue unchanged (base case) or allowing a transforming action to return a modified queue.  Actions are processed in their original order, and if an item is added during this process, it does not have a chance to transform - same as iOS.